src/build/__init__.py

```diff
@@ -170,8 +199,10 @@ def __init__(
         pyproject_toml_path = os.path.join(source_dir, 'pyproject.toml')
         self._build_system = _parse_build_system_table(_read_pyproject_toml(pyproject_toml_path))
 
+        self.project_name: str | None = _parse_project_name(_read_pyproject_toml(pyproject_toml_path))
         self._backend = self._build_system['build-backend']
 
+        self._requires_for_build_cache: dict[str, set[str] | None] = {'wheel': None, 'sdist': None}
```

Comment:
I really don't want us having to deal with caching issues - if you wanna "cache" the result you can save it to a variable.  If it is imperative that the output of `requires_for_build_x` is cached, you should have to opt into reading the requires from the cache.

Reply:
I recall I had implemented this caching to try and avoid calling `get_requires_for_build_wheel` or `get_requires_for_build_sdist` more than once for a build. I'm thinking this caching behavior probably isn't actually necessary if I were to just re-query `get_requires_for_build_{distribution}` a second time. I'll try and rework this.

Reply:
I removed the caching which it appears I had implemented to avoid log message changes breaking some other tests, I tweaked the tests and used a different approach which seems a bit simpler.